src/mlib/int128.test.cpp

```diff
@@ -0,0 +1,305 @@
+   CHECK (two == mlib_int128_add (two, zero));
+
+   // Addition wraps:
+   mlib_int128 max = MLIB_INT128_SMAX;
+   auto more = mlib_int128_add (max, four);
+   CHECK (more == mlib_int128_add (MLIB_INT128_SMIN, MLIB_INT128 (3)));
+
+   // "Wrap" around zero:
+   auto ntwo = MLIB_INT128_CAST (-2);
+   auto sum = mlib_int128_add (ntwo, four);
+   CHECK (sum == two);
+
+   auto eight = mlib_int128_lshift (two, 2);
+   CHECK (eight == MLIB_INT128 (8));
+
+   auto big = mlib_int128_lshift (two, 72);
+   CHECK (mlib_int128_scmp (big, MLIB_INT128 (0)) > 0);
+
+   auto four_v2 = mlib_int128_lshift (eight, -1);
+   CHECK (four == four_v2);
+
+   auto r = mlib_int128_divmod (27828649044156246570177174673037165454_i128,
+                                499242349997913298655486252455941907_i128);
+
+   CHECK (r.quotient == 55_i128);
+   CHECK (r.remainder == 370319794271015144125430787960360569_i128);
+
+   // Self-divide:
+   r = mlib_int128_divmod (628698094597401606590302208_i128,
+                           628698094597401606590302208_i128);
+   CHECK (r.quotient == 1_i128);
+   CHECK (r.remainder == 0_i128);
+
+   // With no high-32 bits in the denominator
+   r = mlib_int128_divmod (316356263640858117670580590964547584140_i128,
+                           13463362962560749016052695684_i128);
+   CHECK (r.quotient == 23497566285_i128);
+
+   // Remainder correctness with high bit set:
+   auto rem = mlib_int128_mod (292590981272581782572061492191999425232_i128,
+                               221673222198185508195462959065350495048_i128);
+   CHECK (rem == 70917759074396274376598533126648930184_i128);
+
+   // Remainder with 64bit denom:
+   rem = mlib_int128_mod (2795722437127403543495742528_i128, 708945413_i128);
+   CHECK (rem == 619266642_i128);
+
+   // 10-div:
+   CHECK (mlib_int128_div (MLIB_INT128_SMAX, 10_i128) ==
+          17014118346046923173168730371588410572_i128);
+
+   int8_t n = -12;
+   CHECK (mlib_int128_scmp (zero, MLIB_INT128_CAST (n)) > 0);
+   CHECK (mlib_int128_ucmp (zero, MLIB_INT128_CAST (n)) < 0);
+
+   auto _2pow127 = mlib_int128_pow2 (127);
+   CHECK (std::string (mlib_int128_format (_2pow127).str) ==
+          "170141183460469231731687303715884105728");
+
+   for (int i = 0; i < 100000; ++i) {
+      std::uniform_int_distribution<std::uint64_t> dist;
+      std::random_device r;
+      auto bits = r ();
+      mlib_int128 num = MLIB_INT128_FROM_PARTS (bits & 0b0001 * dist (r),
+                                                bits & 0b0010 * dist (r));
+      mlib_int128 denom = MLIB_INT128_FROM_PARTS (bits & 0b0100 * dist (r),
+                                                  bits & 0b1000 * dist (r));
+      if (mlib_int128_eq (denom, 0_i128)) {
+         continue;
+      }
+
+      // Attempt a division, even if we have no "reference" value
+      auto q = mlib_int128_divmod (num, denom);
+      (void) q;
+
+#ifdef __SIZEOF_INT128__
+      // When we have an existing i128 impl, test against that:
```

Comment:
Neat idea. Consider fuzz testing division, and comparing the result with the native support. That may help to explore more code paths.